eras/mary/impl/testlib/Test/Cardano/Ledger/Mary/Arbitrary.hs

```diff
@@ -89,9 +93,52 @@ instance Crypto c => Arbitrary (PolicyID c) where
         , elements hashOfDigitByteStrings
         ]
 
+genMultiAssetTriple :: Crypto c => Gen Int64 -> Gen (PolicyID c, AssetName, Int64)
+genMultiAssetTriple genAmount = (,,) <$> arbitrary <*> arbitrary <*> genAmount
+
+-- | The number 910 is chosen with the following rationale.
+--
+-- When we generate a number of MultiAssets all at once, that number happens to have
+-- an implicit upper limit due to the Cardano.Ledger.Mary.Value.{to,from}-based
+-- compacting operation. This operation is also performed when we serialise to and from CBOR.
+--
+-- Refering to the haddock for 'Cardano.Ledger.Mary.Value.to' we surmise that
+--   1. The offsets for AssetName and PolicyID are stored as Word16 (maxBound = 65535).
+--   2. All offsets (including those for AssetName and PolicyID) are relative to the whole
+--      of the representation (binary blob) rather than the start of their respective regions.
+--   3. If the offsets exceed their maxBounds, they will overflow.
+--   4. So, we need to ensure that at least the last of the offsets (AssetName offsets) do
+--      not exceed 65535.
+--   5. With `n` as the total number of assets, the inequality to be satisfied is thus:
+--           8n -- Word64 asset quantities
+--        +  2n -- Word16 policy id offsets
+--        +  2n -- Word16 asset name offsets
+--        + 28n -- 28-byte policy ids
+--        + 32n -- 32-byte asset names (a maximum of 32 bytes)
+--        should be <= 65535
+--        65535 / 72 ~ 910.2 is the maximum number of triples to be safely generated
+--
+-- NOTE: There are some conditions due to which exceeding this number may not
+-- result in a guaranteed failure to compact without overflow, because, during compacting
+--   1. The asset names and policy ids are deduplicated
+--   2. Not all generated asset names are 32-bytes long
+-- But, exceeding this number does make the probability of causing overflow > 0.
 genMultiAsset :: Crypto c => Gen Integer -> Gen (MultiAsset c)
 genMultiAsset genAmount =
-  MultiAsset <$> genNonEmptyMap arbitrary (genNonEmptyMap arbitrary genAmount)
+  multiAssetFromListBounded
+    <$> resize 910 (listOf $ genMultiAssetTriple $ fromIntegral <$> genAmount)
```

Comment:
> This is not gonna work, because PolicyIDs are generated uniformly. This means all of them would only have a single AssetName.

:facepalm: I didn't think of this! :exploding_head: Thanks!

> Also generating asset names on the order of hundreds as a common case is going to destroy the runtime and is guaranteed to be killed with OOM on some serialization roundtrippping, like Block for example.

:+1: :100: :memo: